Python binding to export bytecode format for lite interpreter

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32621 Python binding to export bytecode format for lite interpreter**

Export the "_save_for_mobile" method to Python so that the bytecode format for lite interpreter can be added or updated to the original script model. 

It's the first step of python binding for lite interpreter, as discussed in this [internal post](https://fb.workplace.com/groups/1144215345733672/permalink/1478900738931796/) and offline. 

Next step is to export the load_for_mobile and run method of mobile module, so that users could verify the mobile model from Python. 

Test: use the following python script to display the bytecode part of the updated model file. 
```
#!/usr/bin/env python3
import sys
import pickle
import pprint
import zipfile


class FakeObject(object):
    def __init__(self, module, name, args):
        self.module = module
        self.name = name
        self.args = args
        self.state = None

    def __repr__(self):
        state_str = "" if self.state is None else f"(state={self.state!r})"
        return f"{self.module}.{self.name}{self.args!r}{state_str}"

    def __setstate__(self, state):
        self.state = state


class FakeClass(object):
    def __init__(self, module, name):
        self.module = module
        self.name = name
        self.__new__ = self.fake_new

    def __repr__(self):
        return f"{self.module}.{self.name}"

    def __call__(self, *args):
        return FakeObject(self.module, self.name, args)

    def fake_new(self, *args):
        return FakeObject(self.module, self.name, args)


class DumpUnpickler(pickle._Unpickler):
    def find_class(self, module, name):
        return FakeClass(module, name)

    def persistent_load(self, pid):
        return FakeObject("pers", "obj", (pid,))


def main(argv):
    zfile = zipfile.ZipFile(argv[1])
    names = [i for i in zfile.namelist() if "bytecode.pkl" in i]
    if not names: 
        print("bytecode.pkl not found.")
        return
    with zfile.open(names[0], "r") as handle:
        value = DumpUnpickler(handle).load()
    pprint.pprint(value)


if __name__ == "__main__":
    sys.exit(main(sys.argv))

```

Differential Revision: [D19596359](https://our.internmc.facebook.com/intern/diff/D19596359)

Commit:
Python binding for lite interpreter

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f8d06def:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=d2bda53f6d6cbdf0b2370b769844b4ab71bba7a2&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base d2bda53f since Jan 24
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  Since your merge base is older than `viable/strict`, run these commands:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f8d06def29fe8bc3ddb64a98389bf671af791da2).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=90883434) from Jan 24 until Jan 24



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "Python binding for lite interpreter"

Commit:
Update on "Python binding for lite interpreter"

Commit:
Update on "Python binding to export bytecode format for lite interpreter"


Export the "_save_for_mobile" method to Python so that the bytecode format for lite interpreter can be added or updated to the original script model. 

It's the first step of python binding for lite interpreter, as discussed in this [internal post](https://fb.workplace.com/groups/1144215345733672/permalink/1478900738931796/) and offline. 

Next step is to export the load_for_mobile and run method of mobile module, so that users could verify the mobile model from Python. 

Test: use the following python script to display the bytecode part of the updated model file. 
```
#!/usr/bin/env python3
import sys
import pickle
import pprint
import zipfile


class FakeObject(object):
    def __init__(self, module, name, args):
        self.module = module
        self.name = name
        self.args = args
        self.state = None

    def __repr__(self):
        state_str = "" if self.state is None else f"(state={self.state!r})"
        return f"{self.module}.{self.name}{self.args!r}{state_str}"

    def __setstate__(self, state):
        self.state = state


class FakeClass(object):
    def __init__(self, module, name):
        self.module = module
        self.name = name
        self.__new__ = self.fake_new

    def __repr__(self):
        return f"{self.module}.{self.name}"

    def __call__(self, *args):
        return FakeObject(self.module, self.name, args)

    def fake_new(self, *args):
        return FakeObject(self.module, self.name, args)


class DumpUnpickler(pickle._Unpickler):
    def find_class(self, module, name):
        return FakeClass(module, name)

    def persistent_load(self, pid):
        return FakeObject("pers", "obj", (pid,))


def main(argv):
    zfile = zipfile.ZipFile(argv[1])
    names = [i for i in zfile.namelist() if "bytecode.pkl" in i]
    if not names: 
        print("bytecode.pkl not found.")
        return
    with zfile.open(names[0], "r") as handle:
        value = DumpUnpickler(handle).load()
    pprint.pprint(value)


if __name__ == "__main__":
    sys.exit(main(sys.argv))

```

suo commented on torch/jit/__init__.py:
```diff
@@ -1653,6 +1653,20 @@ def save(self, *args, **kwargs):
             """
             return self._c.save(*args, **kwargs)
 
+        def save_for_mobile(self, *args, **kwargs):
```
Should this be part of the public API for ScriptModule? I thought we were not exposing the lite interpreter externally, so this may be confusing for external users.

iseeyuan replied:
I thought the "_" is needed in C++ only. I might missed it here. Just to confirm: should we rename it to "_save_for_mobile"? I could have a quick patch for it.
